File: Assignment 3/q7_15.m
% a) Competing Dynamics Model
true_set = [];
for n = 1:5 
    
    x = 3*rand(100,1);
    y = 2*rand(100,1);
    sets = [x,y];
    delta_t = 0.1;

    for col=1:size(sets)
        x = sets(col,1);
        y = sets(col,2);

        x_dot = ((3-x-2*y)*x) + delta_t*((3-x-2*y)*x);
        y_dot = ((3-x-2*y)*x) + delta_t*((2-y-x)*y);
        new_set(col,:) = [x_dot,y_dot];
    end
    figure
    hold on;
    plot(new_set(:,1),new_set(:,2));
    hold on;
%     true_set = [true_set, new_set];
%     s = new_set(1)
%     true_set = ([new_set(:,1),new_set(:,2)]);
%     m2 = m([1,2,4],:); % Extract rows 1, 2, and 4;
%     m2 = m2(:, [1, 2, 4, 6]); Now extract columns 1,2,4,& 6

end


figure;
hold on;
% for v = 1:2:200
% % for col=1:100
%     plot(true_set(:,v), true_set(:,v+1));
% end
% h = cellfun(@(x) plot(x), true_set);
% scatter(true_set(:,1), true_set(:,2));
xlabel('x_dot');
ylabel('y_dot');
grid;
axis equal;

% fplot(x_dot,y_dot);

% Discretizing into discrete time using t_s = 0.1
% Hd = c2d(x_dot,0.1,'foh');


% % b) Classic Predator-Prey Model 
% x = rand(100,100);
% y = rand(100,100);
% theta = 2;
% x_dot = (1-y).*x;
% y_dot = theta.*(x-1).*y;
% 
% % NOT SURE IF THIS IS RIGHT...
% figure;
% % quiver(x, y, x_dot, y_dot);
% plot(x_dot, y_dot);
% xlabel('x_dot');
% ylabel('y_dot');
% grid;
% axis equal;
% 
% 
% % c) Selkov Model 
% x = rand(100,100);
% y = rand(100,100);
% theta = 1;
% x_dot = -x + 0.1.*y + y.*x.^2;
% y_dot = theta - 0.1.*y - y.*x.^2;

File: Assignment 3/q7_15.py
import random
import matplotlib
import matplotlib.pyplot as plot

del_t = 0.1
A = 1

# Function: Plot 
def plotGraph(sets, title, axes, fixedPoints):
    # Draw the lines
    collections = matplotlib.collections.LineCollection(sets)

    # Generate plot
    plot.gca().add_collection(collections)

    # Plot the fixed points for reference
    for x in fixedPoints: 
        plot.plot(x[0], x[1], marker='o', markersize=5, color="red")

    # Print the plot 
    plot.title(title)
    plot.xlabel("x")
    plot.ylabel("y")
    plot.axis([axes[0], axes[1], axes[2], axes[3]])
    plot.draw()
    plot.show()


# Part A - Competing Dynamics Model"
# Create random points
points = [(3*random.uniform(0,1), 2*random.uniform(0,1)) for i in range(0,100)]
sets = [[pt] + 100*[None] for idx, pt in enumerate(points)] # Do this 100 times (as per question)

for i, pt_1 in enumerate(sets):
    for j in range(0, len(sets[i])):

        x = sets[i][j][0]
        y = sets[i][j][1]
        x_dot = (3-x-2*y)*x
        y_dot = (2-y-x)*y

        # Calculate new x and y values (via Forward Euler)
        new_x = x + del_t*A*x_dot 
        new_y = y + del_t*A*y_dot

        # Replace the new discretized values to the next point 
        if (j+1 < len(sets[i])):            
            sets[i][j+1] = (new_x, new_y)
fixedPoints = [[0,0], [0,2], [3,0], [1,1]]
plotGraph(sets, "Part A - Competing Dynamics Model", [-0.5,3.5,-0.5,2.5], fixedPoints)

# Part B - Classic Predator–Prey Model
# Create random points 
points_b = [(random.uniform(0,1), random.uniform(0,1)) for i in range(0,100)]
sets_b = [[pt] + 100*[None] for idx, pt in enumerate(points_b)] # Do this 100 times (as per question)
theta = 2

for i, pt_1 in enumerate(sets_b):
    for j in range(0, len(sets_b[i])):

        x = sets_b[i][j][0]
        y = sets_b[i][j][1]
        x_dot = (1-y)*x
        y_dot = theta*(x-1)*y

        # Calculate new x and y values (via Forward Euler)
        new_x = x + del_t*A*x_dot 
        new_y = y + del_t*A*y_dot

        # Replace the new discretized values to the next point 
        if (j+1 < len(sets_b[i])):            
            sets_b[i][j+1] = (new_x, new_y)

fixedPoints = [[0,0], [1,1]]
plotGraph(sets_b, "Part B - Classic Predator–Prey Model", [0,6,0,6], fixedPoints)

# Part C - Selkov Model
# Create random points 
points_c = [(random.uniform(0,1), random.uniform(0,1)) for i in range(0,5)]
sets_c = [[pt] + 100*[None] for idx, pt in enumerate(points_c)] # Do this 100 times (as per question)
thetas = [0.1, 0.5, 1]

# Do this for Theta = 0.1, 0.5, and 1
for theta in thetas: 
    for i, pt_1 in enumerate(sets_c):
        for j in range(0, len(sets_c[i])):

            x = sets_c[i][j][0]
            y = sets_c[i][j][1]
            x_dot = -x+0.1*y+(x**2)*y
            y_dot = theta-0.1*y-(x**2)*y

            # Calculate new x and y values (via Forward Euler)
            new_x = x + del_t*A*x_dot 
            new_y = y + del_t*A*y_dot

            # Replace the new discretized values to the next point 
            if (j+1 < len(sets_c[i])):            
                sets_c[i][j+1] = (new_x, new_y)

    fixedPoints=[]
    plotGraph(sets_c, "Part C - Selkov Model with theta =  %s" % theta, [0,3,0,3], fixedPoints)

